### PR TITLE
Override completion must not include parameter attributes

### LINKED ISSUE #5646: Override completion includes internal attributes when overriding in a different assembly
We should probably only copy attributes if they are accessible.

### PR DESCRIPTION
**Pull request by ivanbasov**
Fixes #5646

### CODE REVIEW

File: src/Workspaces/Core/Portable/Shared/Extensions/SyntaxGeneratorExtensions.cs [hunk 1]
```diff
@@ -526,6 +526,9 @@ private static DeclarationModifiers GetOverrideModifiers(ISymbol symbol)
                     method: overriddenMethod,
                     accessibility: overriddenMethod.ComputeResultantAccessibility(newContainingType),
                     modifiers: modifiers,
+                    // There is no need for an override to have same parameters attributes as the original method.
+                    // Just clear attributes when generating an override.
+                    parameters: overriddenMethod.Parameters.SelectAsArray(p => p.WithAttributes(ImmutableArray<AttributeData>.Empty)),
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/SyntaxGeneratorExtensions.cs]: what about overridden properties in VB?  they can have parameters too :)

**Comment by genlu** [comment]
Would missing the attribute in overriding method cause anything unpected if the overridden method is abstract?

File: src/Workspaces/Core/Portable/Shared/Extensions/SyntaxGeneratorExtensions.cs [hunk 2]
```diff
@@ -526,6 +526,9 @@ private static DeclarationModifiers GetOverrideModifiers(ISymbol symbol)
                     method: overriddenMethod,
                     accessibility: overriddenMethod.ComputeResultantAccessibility(newContainingType),
                     modifiers: modifiers,
+                    // There is no need for an override to have same parameters attributes as the original method.
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Shared/Extensions/SyntaxGeneratorExtensions.cs]: I worry this is a bit strong of a statement that there's _no_ need -- how do we know there aren't some reflection-based systems out there that are depending on this?

Which attributes triggered this original question?

**Comment by CyrusNajmabadi** [comment]
>Code seems OK, but worried that this is the right thing to do or if we need a less aggressive hammer.

FWIW, i think this is totally ok.  I recall us doing this in many other scenarios (maybe implement interface) and it really hasn't been a problem.   This feels more like a missed case, versus something that is important to preserve behavior on.

File: src/Workspaces/Core/Portable/CodeGeneration/CodeGenerationSymbolFactory.cs [hunk 3]
```diff
@@ -423,6 +423,7 @@ public static INamespaceSymbol CreateNamespaceSymbol(string name, IList<ISymbol>
             DeclarationModifiers? modifiers = null,
             ImmutableArray<IMethodSymbol> explicitInterfaceImplementations = default,
             string name = null,
+            ImmutableArray<IParameterSymbol>? parameters = default,
```
Comment (genlu) [inline src/Workspaces/Core/Portable/CodeGeneration/CodeGenerationSymbolFactory.cs]: nit:
`ImmutableArray<IParameterSymbol>? parameters = null` just to be consistent with others

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/CodeGeneration/CodeGenerationSymbolFactory.cs]: ImmutableArray is a struct, so we need to use `default` which is also above for `explicitInterfaceImplementations`

Comment (genlu) [reply src/Workspaces/Core/Portable/CodeGeneration/CodeGenerationSymbolFactory.cs]: This is a `Nullable<T>` though :)

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/CodeGeneration/CodeGenerationSymbolFactory.cs]: Touche!  Why is it nullable though?  an ImmutbaleArray can already represent '3' states (null/emty/elements), we do we need an additional Nullable on top of that?

Comment (ivanbasov) [reply src/Workspaces/Core/Portable/CodeGeneration/CodeGenerationSymbolFactory.cs]: 1. ImmutableArray is a struct: default should be used not null.
2. ImmutableArray has enough states to distinguish cases we consider: default, empty, items.

No changes required. Thank you!

---
In reply to: [290923010](https://github.com/dotnet/roslyn/pull/36163#discussion_r290923010) [](ancestors = 290923010)

**Comment by ivanbasov** [comment]
As far as I know, member signatures do not depend on attributes. In some cases, one may want some attributes to be copied or added. However, we do not need to guarantee this. 

If you have an example of the opposite, we definitely should consider it.

---
In reply to: [499193025](https://github.com/dotnet/roslyn/pull/36163#issuecomment-499193025) [](ancestors = 499193025)

File: src/Workspaces/Core/Portable/Shared/Extensions/SyntaxGeneratorExtensions.cs (hunk 2 above)
Comment (ivanbasov) [reply src/Workspaces/Core/Portable/Shared/Extensions/SyntaxGeneratorExtensions.cs]: We already do not copy attributes of methods and attributes of properties. By mistake (my opinion), we copy attributes of method parameters and property parameters. We just should stop doing this.

---
In reply to: [290869693](https://github.com/dotnet/roslyn/pull/36163#discussion_r290869693) [](ancestors = 290869693)

**Comment by ivanbasov** [comment]
- support of property parameters added
- feedback responded

@CyrusNajmabadi , @genlu , @jasonmalinowski  please review

**Comment by ivanbasov** [comment]
There is a test failing: `TestPropertyReturnTypeAttributes`
The test was added by @CyrusNajmabadi to fix http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/546779
saying if we copy parameter attributes for properties, we also need to copy return type attributes for properties in case scenario:
```
Imports System.Runtime.InteropServices
 
Interface I
    Property P(<MarshalAs(UnmanagedType.I4)> x As Integer) As <MarshalAs(UnmanagedType.I4)> Integer
End Interface
 
Class C
    Implements I ' Implement
End Class
```
This is the concern that @genlu and @jasonmalinowski mentioned. The question is: do we need to copy attributes for any purpose? I'd say - no. See ideas in comments above. @CyrusNajmabadi , @jasonmalinowski , @genlu - your thoughts?

**Comment by jasonmalinowski** [comment]
The COM marshalling case does seem worrisome: I would be most surprised if the CLR walked to the base type to find the marshalling attributes before knowing how to marshal things. And in that case, those attributes are 100% critical to the correct functioning of the method.

More generally, the only way this is safe to blanket apply to all methods is if we know that _every framework_ using reflection or metadata walking is going to look to the base type. That's a huge proof I don't think we can possibly make.

Looking at the other bug, if the concern is NullableAttribute then we should close this PR and dupe it against one of the work items for nullable, since we need to address that attribute holistically and we're still having conversations about the best way to do it. In any case, this PR _isn't_ the right way to fix that. If some other attribute is causing us to look at this then let's discuss _that attribute_ -- but otherwise I'd absolutely close this PR and by design the original bug, making sure we have a tracking bug for NullableAttribute as appropriate.

**Comment by ivanbasov** [comment]
Synced with @jasonmalinowski offline. Agreed to keep accessible attributes only. Decided to add a method to determine attributes to be kept vs attributes to be dropped. @jasonmalinowski  will extend it for other purposes later.

@jasonmalinowski , @genlu , @CyrusNajmabadi please review.

**Pull request merged**